src/stream/src/from_proto/chain.rs

```diff
@@ -43,26 +50,98 @@ impl ExecutorBuilder for ChainExecutorBuilder {
         // its schema.
         let schema = snapshot.schema().clone();
 
-        if node.disable_rearrange {
-            let executor = ChainExecutor::new(
+        let executor = match node.chain_type() {
+            ChainType::Chain => ChainExecutor::new(
                 snapshot,
                 mview,
                 upstream_indices,
                 progress,
                 schema,
                 params.pk_indices,
-            );
-            Ok(executor.boxed())
-        } else {
-            let executor = RearrangedChainExecutor::new(
+            )
+            .boxed(),
+            ChainType::Rearrange => RearrangedChainExecutor::new(
                 snapshot,
                 mview,
                 upstream_indices,
                 progress,
                 schema,
                 params.pk_indices,
-            );
-            Ok(executor.boxed())
-        }
+            )
+            .boxed(),
+            ChainType::Backfill => {
+                let table_desc: &StorageTableDesc = node.get_table_desc()?;
+                let table_id = TableId {
+                    table_id: table_desc.table_id,
+                };
+
+                let order_types = table_desc
+                    .pk
+                    .iter()
+                    .map(|desc| {
+                        OrderType::from_prost(&ProstOrderType::from_i32(desc.order_type).unwrap())
+                    })
+                    .collect_vec();
+
+                let column_descs = table_desc
+                    .columns
+                    .iter()
+                    .map(ColumnDesc::from)
+                    .collect_vec();
+                let column_ids = column_descs.iter().map(|x| x.column_id).collect_vec();
+
+                // Use indices based on full table instead of streaming executor output.
+                let pk_indices = table_desc.pk.iter().map(|k| k.index as usize).collect_vec();
+
+                let dist_key_indices = table_desc
+                    .dist_key_indices
+                    .iter()
+                    .map(|&k| k as usize)
+                    .collect_vec();
+                let distribution = match params.vnode_bitmap {
+                    Some(vnodes) => Distribution {
+                        dist_key_indices,
+                        vnodes: vnodes.into(),
+                    },
+                    None => Distribution::fallback(),
+                };
+
+                let table_option = TableOption {
+                    retention_seconds: if table_desc.retention_seconds > 0 {
+                        Some(table_desc.retention_seconds)
+                    } else {
+                        None
+                    },
+                };
+                let value_indices = table_desc
+                    .get_value_indices()
+                    .iter()
+                    .map(|&k| k as usize)
+                    .collect_vec();
+                let table = StorageTable::new_partial(
```

Comment:
Can we refactor this with `from_table_catalog` in the future?

Reply:
Sure, I will add a TODO here.